### ISSUE TITLE
preprocessing for VGG input

### ISSUE THREAD

**Issue by ruotianluo** [tensorflow/models#517]
Hi, I want to use the pretrained VGG to do other task.
I saw in other's vgg code, there is rgb to bgr transform. I don't see this transform in preprocessing in the models/slim/preprocessing/vgg_preprocessing.py.

Should there be rgb2bgr in preprocessing?

**Comment by tatatodd** [comment]
@ruotianluo We primarily use github issues to track bugs and feature requests.  This is a question better suited for StackOverflow, which we also monitor.  Please ask it there and tag it with the `tensorflow` tag.  Thanks!

**Comment by ruotianluo** [comment]
@tatatodd I am reporting a bug which I'm not sure about if it is.

**Comment by ruotianluo** [comment]
@tatatodd

**Comment by ruotianluo** [comment]
@tatatodd Can I know if the vgg checkpoint you provide is converted from the original caffe-model? If it's converted, did you change the channel order in the first conv layer.

**Comment by tatatodd** [comment]
@ruotianluo I see.

@nathansilberman might be able to answer your question about whether the channels are rgb or bgr.

**Comment by Hediby** [comment]
I am also interested in the answer

**Comment by ruotianluo** [comment]
I tried to follow the way used in magenta style transfer where they also didn't shift the channels.

And I compared fc7 and fc8 feature of the same image from slim and caffe. The results are very different. However, even I shift the channels, there is still some gap. Is the vgg model trained from scratch, or copied from the caffe model?

**Comment by joncox123** [comment]
I am having what seems to be related issues that only occur with models trained on ImageNet. I have tried using stock TF-Slim and writing all of my own custom TensorFlow code. Also, I have tried using Google's VGG-16 checkpoint, as well as writing my own converter from Caffe models to TensorFlow. 

When I input the exact same data in Caffe and TensorFlow, I get virtually identical outputs for each layer for both tools. Of course, I do the comparison after the necessary tensor reordering. 

However, I am still getting incorrect results with TensorFlow, which suggests a problem with input or output (label) data. 

**I assume this is a stupid question, but has anybody had success getting VGG-16 to work on ImageNet 2012 using stock TF Slim?** 

With my code, I am computing the softmax activation and cross entropy loss myself, then the top 1 and top 5 accuracy. This works fine with MNIST and CIFAR-10, so I don't believe I have a bug here.

**Comment by heshameraqi** [comment]
I am also interested in an answer.

**Comment by heshameraqi** [comment]
@ruotianluo Maybe they are different because of dropouts. Did you disable them ?

**Comment by MarcelSimon** [comment]
The missing explanation in the source code certainly can be considered a bug. There should be a specification of the input data

**Comment by wookayin** [comment]
So anybody can please clarify whether the input to `vgg_16` or `vgg_19` be

- (i) in float32 [0, 1] scale or [0, 255] scale
- (ii) with VGG mean pixels subtracted (I think so)
- (iii) in RGB order or BGR order?

There is no documentation or specification on slim models and pre-trained checkpoints ...

**Comment by simo23** [comment]
Hi, in my tests I use vgg16 and vgg19 from slim and I do the following:

            # Read the image from file
            image_string = tf.read_file(filename)
            image_decoded = tf.image.decode_jpeg(image_string, channels=3)    # The decoded image is in RGB format
            image = tf.cast(image_decoded, tf.float32)

            # Isotropic scaling
            smallest_side = 256.0
            height, width = tf.shape(image)[0], tf.shape(image)[1]
            height = tf.to_float(height)
            width = tf.to_float(width)
            scale = tf.cond(tf.greater(height, width),lambda:smallest_side/width,lambda:smallest_side/height)
            new_height = tf.to_int32(height * scale)
            new_width = tf.to_int32(width * scale)
            image = tf.image.resize_images(image, [new_height, new_width]) 

            VGG_MEAN = [123.68, 116.78, 103.94]    # This is R-G-B for Imagenet

            image = tf.random_crop(image, [224, 224, 3])  
            means = tf.reshape(tf.constant(VGG_MEAN), [1, 1, 3])
            image = image - means 

The input images are in range [0,255] read in RGB format, with the RGB means from Imagenet subtracted. My results are fine so I think this preprocessing is correct. 

Hope it helps!

**Comment by zgongkuang** [comment]
@simo23, Thanks for your example. I always have one question. Why we should substruct the training mean, not the testing mean? For example, if my input test image has mean value much lower than 103.94, though in the scale of [0 255],  should I still use the Imagenet training mean?  I notice you use rescaling to set the image size to 256, then truncate to 224. My input size is 128, can I padding it with zeros and  enlarge it into 224?  Or rescaling it from 128 to 256, then truncate to 224 is better?

**Comment by simo23** [comment]
Hi, I m sure there are people with a 1M times better answer than me but I'll give it a try:

- The VGG standard mean has been obtained by zero centering the distribution of the whole training set of Imagenet. This is usually done because it is proved to improve accuracy and help the learning process. 

- We do not subtract the testing mean because we should not know the testing mean, you should only rely on the training data that you have. 

- We do not subtract each image's mean because if we subtract the official VGG mean then the features that will be extracted by the network will be the "standard" VGG ones which are proved to achieve a certain accuracy. If you subtract each image mean then the feature extracted by the network will not be the "standard" ones so your results probably will not be the same. As a practical example: if you have two images, one totally red and one totally blue, by subtracting the single image mean they will result equal after preprocessing and we do not want that. So we do not subtract the single image mean to image itself but by a mean that comes from a distribution that represents the whole dataset. 

For your other more practical questions, I've not experimented anything like that before so I can only guess:

- If your input size is 128 you can still put the image in input. You will obtain a 4x4 feature vector after the last conv5 layer which I don't think will be enough to be fed into a fully connected. Zero padding the image to be 224 should not improve your performances as the data is still a 128x128. 

- You can try to resize the image to 256 by zooming it and the apply random sampling to get the 224x224. This to me sound better than feed the 128 image, but you have to try. 

Hope it helps!

**Comment by karmel** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by neouyghur** [comment]
It is RGB. Please read the comments in this file. https://github.com/tensorflow/models/blob/master/research/slim/datasets/build_imagenet_data.py
This should be included in Slim manual.

**Comment by alexanderkoumis** [comment]
Sorry if I am bumping an old issue, but there is a highly voted comment that says the image format that worked for them was BGR and then a followup comment with links to the docs that says RGB. Which is it?

**Comment by jiangyangzhou** [comment]
I think it's a bug of Readme, Cause you can't use pretrain model in Slim without pretrain information.

**Comment by jiangyangzhou** [comment]
Best Ans: https://github.com/tensorflow/models/blob/master/research/slim/preprocessing/vgg_preprocessing.py
Only subtract RGB value.
